src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -1069,6 +1069,12 @@ private EnumeratorResult SatisfiesIEnumerableInterfaces(SyntaxNode collectionSyn
             Debug.Assert((object)builder.CollectionType != null);
 
             NamedTypeSymbol collectionType = (NamedTypeSymbol)builder.CollectionType;
+
+            if (unwrappedCollectionExprType.IsRefLikeType || unwrappedCollectionExprType is TypeParameterSymbol { AllowByRefLike: true })
+            {
+                builder.CollectionType = unwrappedCollectionExprType;
```

Comment:
>builder.CollectionType [](http://example.com/codeflow?start=16&length=22)

Should the spec-let update the definition of *collection type* for `ref struct` and `allows ref struct` cases?

Reply:
> > builder.CollectionType
> 
> Should the spec-let update the definition of _collection type_ for `ref struct` and `allows ref struct` cases?

Could you elaborate please? What are you suggesting to modify and where?

Reply:
Should https://github.com/dotnet/csharplang/pull/8024 include a section that updates the [*foreach statement*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) definition to indicate the expected *collection type*, etc. for `ref struct` cases and type parameters with `allows ref struct`? Currently, the existing spec section seems to indicate the *collection type* will be `IEnumerable<T>` or `IEnumerable`. Are those correct, given how *collection type*, etc. is used in the `foreach` expansion and perhaps elsewhere?

Reply:
> Should [dotnet/csharplang#8024](https://github.com/dotnet/csharplang/pull/8024) include a section that updates the [_foreach statement_](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) definition to indicate the expected _collection type_, etc. for `ref struct` cases and type parameters with `allows ref struct`? Currently, the existing spec section seems to indicate the _collection type_ will be `IEnumerable<T>` or `IEnumerable`. Are those correct, given how _collection type_, etc. is used in the `foreach` expansion and perhaps elsewhere?

I am satisfied with how  [dotnet/csharplang#8024](https://github.com/dotnet/csharplang/pull/8024) specifies the expected behavior. I will mention that [_foreach statement_](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) will have to be adjusted accordingly. However, I prefer leaving this exercise for later and quite possibly for the one changing the official spec.